lib/prism/parse_result.rb

```diff
@@ -221,6 +217,13 @@ def join(other)
     def self.null
       new(nil, 0, 0)
     end
+
+    private
+
+    # Access the associated source, raising if not present.
+    def source!
```

Comment:
Why is this necessary? Shouldn't a source always be present?

Reply:
See discussion with @eregon at https://github.com/ruby/prism/pull/2101#discussion_r1445829919

Because of the implementations of `Location.null` and `DSL#Location`, Location can have a nil source. We can switch back to saying Location always has a source and then ignoring type-check errors in the other places. What would you prefer?

Reply:
That's a great suggestion, let's ignore `Location.null` and `DSL#Location`, then type this as always being present.

Reply:
I want to clarify the effects of marking as always non-nil:

Marking it as non-nil is unsound (as mentioned in the linked thread) and will mean NoMethodError instead of a clear(er) error when using Location methods that need a source.

For example here are two bugs on current master that correct typing would catch and improve:
```
$ bundle exec ruby -Ilib -rprism -e 'p Prism::Location.null'
lib/prism/parse_result.rb:207:in `start_line': undefined method `line' for nil:NilClass (NoMethodError)

      source.line(start_offset)
            ^^^^^
	from /home/eregon/code/yarp/lib/prism/parse_result.rb:170:in `inspect'
	from -e:1:in `p'
	from -e:1:in `<main>'

$ bundle exec ruby -Ilib -rprism -e 'p Prism::Location.null.start_character_offset'
lib/prism/parse_result.rb:181:in `start_character_offset': undefined method `character_offset' for nil:NilClass (NoMethodError)

      source.character_offset(start_offset)
            ^^^^^^^^^^^^^^^^^
	from -e:1:in `<main>'
```
compared to on this branch:
```
$ be ruby -Ilib -rprism -e 'p Prism::Location.null'                       
lib/prism/parse_result.rb:291:in `source!': Missing source (RuntimeError)
	from /home/eregon/code/yarp/lib/prism/parse_result.rb:203:in `start_line'
	from /home/eregon/code/yarp/lib/prism/parse_result.rb:166:in `inspect'
	from -e:1:in `p'
	from -e:1:in `<main>'
```

`Location.null` doesn't seem used anymore, so maybe that could be removed (OTOH that seems more efficient than many "empty" Location instances).
But `DSL#Location` is used.
One thought would be to use `nil` instead of a "location without source" but then that would make all location fields nil-able which seems way worse.
Another possibility would be to make `DSL#Location` use a "dummy/not-provided/empty" Source as default when not passed a Source. Maybe the best alternative?

Changing things more seems out of scope of this PR, so I think it's OK to merge with forcing the type as non-nil (or with the current state with `source!`, IMO that's better but @kddnewton decides).
It seems something worth solving properly though, as e.g. unsound types can be worse than no types in some aspects.

Reply:
Maybe this is a documentation problem then. `Prism::Location.null` and `DSL::Location` are meant to provide a common interface, but not necessary to actually be used. I've used them both in testing, which is why they're in there in the first place. Honestly I've been thinking about dumping DSL for a while now anyway, and I know `Prism::Location.null` isn't actually used. Since these are lower-priority rarely used things, I don't want them to "infect" the types of other things.

Reply:
Makes sense, let's solve this later then.

The DSL seems to mainly make things slightly shorter and an optional location:
```ruby
    def AliasMethodNode(new_name, old_name, keyword_loc, source = nil, location = Location())
      AliasMethodNode.new(source, new_name, old_name, keyword_loc, location)
    end
```
I think it's quite nice in `test/prism/errors_test.rb` though and would be useful for gems that transform the AST, e.g. to do their own custom desugaring or anything else.

I think the best and simplest solution there would be to use an empty source, like:
```ruby
    EMPTY_SOURCE = Source.new("")
    EMPTY_LOCATION = Location.new(EMPTY_SOURCE, 0, 0)

    def Location(source, start_offset, length)
      Location.new(source, start_offset, length)
    end

    def AliasMethodNode(new_name, old_name, keyword_loc, source = EMPTY_SOURCE, location = EMPTY_LOCATION)
      AliasMethodNode.new(source, new_name, old_name, keyword_loc, location)
    end
```

Reply:
I went with @kddnewton's preferred approach in 2d859af588500f4a708f6415feeab6c56b85bcf8, but I like @eregon's null object pattern approach and will throw up in a later PR to see what you guys think.

While I have you, how do you feel about committing templated codegen'ed output files into the repository? I'd add some sort of lint or test to ensure there's no drift.

It would ease browsing code on GithHub or comparing codegen'ed files between branches locally. But I understand if you don't prefer that.